File: components/esp_board_manager/devices/dev_led_strip/dev_led_strip.c
/*
 * SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO., LTD
 * SPDX-License-Identifier: LicenseRef-Espressif-Modified-MIT
 *
 * See LICENSE file for details.
 */

#include <inttypes.h>
#include "esp_log.h"
#include "esp_err.h"
#include "esp_board_entry.h"
#include "esp_board_device.h"
#include "dev_led_strip.h"

static const char *TAG = "DEV_LED_STRIP";

int dev_led_strip_init(void *cfg, int cfg_size, void **device_handle)
{
    if (cfg == NULL || device_handle == NULL) {
        ESP_LOGE(TAG, "Invalid parameters");
        return -1;
    }
    if (cfg_size != sizeof(dev_led_strip_config_t)) {
        ESP_LOGE(TAG, "Invalid config size");
        return -1;
    }

    const dev_led_strip_config_t *config = (const dev_led_strip_config_t *)cfg;
    if (config->sub_type == NULL) {
        ESP_LOGE(TAG, "Invalid LED strip sub_type");
        return -1;
    }

    ESP_LOGI(TAG, "Initializing LED strip: %s, chip: %s, sub_type: %s, gpio: %d, max_leds: %" PRIu32,
             config->name,
             config->chip ? config->chip : "(unknown)",
             config->sub_type, config->strip_config.strip_gpio_num,
             config->strip_config.max_leds);

    const esp_board_entry_desc_t *entry_desc = esp_board_entry_find_subtype_desc("led_strip", config->sub_type);
    if (entry_desc == NULL || entry_desc->init_func == NULL) {
        ESP_LOGE(TAG, "Failed to find LED strip sub device: %s", config->sub_type);
        return -1;
    }

    int ret = entry_desc->init_func(cfg, cfg_size, device_handle);
    if (ret != 0) {
        ESP_LOGE(TAG, "Failed to initialize LED strip sub device: %s", config->sub_type);
        return ret;
    }

    ESP_LOGI(TAG, "Successfully initialized LED strip: %s, sub_type: %s",
             config->name, config->sub_type);
    return 0;
}

int dev_led_strip_deinit(void *device_handle)
{
    if (device_handle == NULL) {
        ESP_LOGE(TAG, "Invalid parameters");
        return -1;
    }

    dev_led_strip_config_t *cfg = NULL;
    esp_board_device_get_config_by_handle(device_handle, (void **)&cfg);
    if (cfg == NULL) {
        ESP_LOGE(TAG, "Failed to get LED strip config by handle");
        return -1;
    }
    if (cfg->sub_type == NULL) {
        ESP_LOGE(TAG, "Invalid LED strip sub_type");
        return -1;
    }

    const esp_board_entry_desc_t *desc = esp_board_entry_find_subtype_desc("led_strip", cfg->sub_type);
    if (desc == NULL || desc->deinit_func == NULL) {
        ESP_LOGE(TAG, "No deinit function found for LED strip sub type: %s", cfg->sub_type);
        return -1;
    }

    int ret = desc->deinit_func(device_handle);
    if (ret != 0) {
        ESP_LOGE(TAG, "LED strip sub type %s deinit failed: %d", cfg->sub_type, ret);
        return ret;
    }

    ESP_LOGI(TAG, "Successfully deinitialized LED strip: %s, sub_type: %s",
             cfg->name, cfg->sub_type);
    return 0;
}

File: components/esp_board_manager/devices/dev_led_strip/dev_led_strip.py
# SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO., LTD
# SPDX-License-Identifier: LicenseRef-Espressif-Modified-MIT
#
# See LICENSE file for details.

# LED strip device config parser
VERSION = 'v1.0.0'

VALID_SUB_TYPES = ['rmt', 'spi']

COLOR_COMPONENT_FORMAT_MAP = {
    'LED_STRIP_COLOR_COMPONENT_FMT_GRB': 'LED_PIXEL_FORMAT_GRB',
    'LED_STRIP_COLOR_COMPONENT_FMT_GRBW': 'LED_PIXEL_FORMAT_GRBW',
}

UNSUPPORTED_COLOR_COMPONENT_FORMATS = {
    'LED_STRIP_COLOR_COMPONENT_FMT_RGB',
    'LED_STRIP_COLOR_COMPONENT_FMT_RGBW',
}


def _parse_pixel_format(config: dict) -> str:
    value = config.get('led_pixel_format', config.get('color_component_format', 'LED_PIXEL_FORMAT_GRB'))
    if value in UNSUPPORTED_COLOR_COMPONENT_FORMATS:
        raise ValueError(
            f'{value} has no exact led_pixel_format equivalent in the current led_strip driver'
        )
    return COLOR_COMPONENT_FORMAT_MAP.get(value, value)


def get_includes() -> list:
    """Return required include headers for LED strip device"""
    return [
        'dev_led_strip.h',
    ]


def _parse_strip_config(config: dict) -> dict:
    """Parse common led_strip_config_t fields"""
    return {
        'strip_gpio_num': int(config.get('strip_gpio_num', -1)),
        'max_leds': int(config.get('max_leds', 1)),
        'led_pixel_format': _parse_pixel_format(config),
        'led_model': config.get('led_model', 'LED_MODEL_WS2812'),
        'flags': {
            'invert_out': config.get('invert_out', config.get('flags', {}).get('invert_out', False)),
        },
    }


def _parse_logical_lights(config: dict, max_leds: int) -> list:
    """Parse optional logical light ranges for one physical LED strip"""
    logical_lights = config.get('logical_lights', [])
    if logical_lights is None:
        logical_lights = []
    if not isinstance(logical_lights, list):
        raise ValueError('logical_lights must be a list')
    if len(logical_lights) > 32:
        raise ValueError('logical_lights supports at most 32 entries')

    parsed = []
    for index, item in enumerate(logical_lights):
        if not isinstance(item, dict):
            raise ValueError(f'logical_lights[{index}] must be a mapping')

        first_led = int(item.get('first_led', item.get('index', index)))
        led_count = int(item.get('led_count', item.get('count', 1)))
        if first_led < 0:
            raise ValueError(f'logical_lights[{index}] first_led must be >= 0')
        if led_count <= 0:
            raise ValueError(f'logical_lights[{index}] led_count must be > 0')
        if first_led >= max_leds or led_count > max_leds - first_led:
            raise ValueError(
                f'logical_lights[{index}] range {first_led}+{led_count} exceeds max_leds {max_leds}'
            )

        parsed.append({
            'id': item.get('id', item.get('name', None)),
            'title': item.get('title', item.get('label', None)),
            'first_led': first_led,
            'led_count': led_count,
        })

    return parsed


def _parse_rmt_sub_config(config: dict) -> dict:
    """Parse led_strip_rmt_config_t fields"""
    rmt_config = config.get('rmt', config.get('rmt_config', {}))
    return {
        'rmt_config': {
            'clk_src': rmt_config.get('clk_src', 'RMT_CLK_SRC_DEFAULT'),
            'resolution_hz': int(rmt_config.get('resolution_hz', 10000000)),
            'mem_block_symbols': int(rmt_config.get('mem_block_symbols', 0)),
            'flags': {
                'with_dma': rmt_config.get('with_dma', rmt_config.get('flags', {}).get('with_dma', False)),
            },
        },
    }


def _parse_spi_sub_config(config: dict) -> dict:
    """Parse led_strip_spi_config_t fields"""
    spi_config = config.get('spi', config.get('spi_config', {}))
    return {
        'spi_config': {
            'clk_src': spi_config.get('clk_src', 'SPI_CLK_SRC_DEFAULT'),
            'spi_bus': spi_config.get('spi_bus', 'SPI2_HOST'),
            'flags': {
                'with_dma': spi_config.get('with_dma', spi_config.get('flags', {}).get('with_dma', True)),
            },
        },
    }


def parse(name: str, full_config: dict, peripherals_dict=None) -> dict:
    """Parse LED strip device configuration from YAML"""
    config = full_config.get('config', {})
    chip = full_config.get('chip', config.get('chip', 'led_strip'))
    sub_type = full_config.get('sub_type')

    if not sub_type:
        raise ValueError(f"LED strip device '{name}' is missing required 'sub_type' field")

    if sub_type not in VALID_SUB_TYPES:
        raise ValueError(
            f"LED strip device '{name}' has invalid 'sub_type' value '{sub_type}'. "
            f'Must be one of: {VALID_SUB_TYPES}'
        )

    strip_config = _parse_strip_config(config)

    if strip_config['strip_gpio_num'] < 0:
        raise ValueError(f"LED strip device '{name}' requires valid strip_gpio_num")
    if strip_config['max_leds'] <= 0:
        raise ValueError(f"LED strip device '{name}' requires max_leds > 0")
    logical_lights = _parse_logical_lights(config, strip_config['max_leds'])

    if sub_type == 'rmt':
        sub_cfg = {'rmt': _parse_rmt_sub_config(config)}
    elif sub_type == 'spi':
        sub_cfg = {'spi': _parse_spi_sub_config(config)}
    else:
        raise ValueError(f'Unsupported LED strip sub_type: {sub_type}')

    struct_init = {
        'name': name,
        'chip': chip,
        'sub_type': sub_type,
        'strip_config': strip_config,
        'logical_light_count': len(logical_lights),
        'sub_cfg': sub_cfg,
    }
    if logical_lights:
        struct_init['logical_lights'] = logical_lights

    return {
        'struct_type': 'dev_led_strip_config_t',
        'struct_var': f'{name}_cfg',
        'struct_init': struct_init,
    }
